src/cargo/ops/cargo_rustc/commands.rs

```diff
@@ -0,0 +1,131 @@
+use std::collections::HashMap;
+use std::c_str::CString;
+use std::io::process::ProcessOutput;
+use std::fmt::{mod, Show, Formatter};
+
+use util::{mod, CargoResult, ProcessError, ProcessBuilder};
+
+/// Trait for objects that can execute commands.
+pub trait CommandsExecution {
+    fn exec(&self, CommandPrototype) -> Result<(), ProcessError>;
+    fn exec_with_output(&self, CommandPrototype) -> Result<ProcessOutput, ProcessError>;
+}
+
+/// Default implementation of `CommandsExecution`.
+#[deriving(Copy)]
+pub struct DefaultCommands;
+
+impl CommandsExecution for DefaultCommands {
+    fn exec(&self, command: CommandPrototype) -> Result<(), ProcessError> {
+        command.into_process_builder().unwrap().exec()
+    }
+
+    fn exec_with_output(&self, command: CommandPrototype)
+                        -> Result<ProcessOutput, ProcessError> {
+        command.into_process_builder().unwrap().exec_with_output()
+    }
+}
+
+/// Prototype for a command that must be executed.
+#[deriving(Clone)]
+pub struct CommandPrototype {
+    ty: CommandType,
+    args: Vec<CString>,
+    env: HashMap<String, Option<CString>>,
+    cwd: Path,
+}
```

Comment:
This does seem to contain an unfortunate amount of duplication with `ProcessBuilder`. Perhaps a `ProcessBuilder` could be stored internally and accessors could be provided?